Use an enum in memory

Eh2406 opened the pull request:
This is an impl of #56.

The main questionable decision is: `remove_last_decision` was handled by removing `add_version` and using `add_decision` directly.  I think the next call to `unit_propagation` will clean up if one of the dependency is satisfied. But I don't know how important that fast path was.

Eh2406 committed:
refactor: use an enum for PackageAssignments

Eh2406 committed:
perf: relation_with with any is a waste

mpizenberg commented on src/internal/memory.rs:
```diff
@@ -92,11 +116,20 @@ impl<P: Package, V: Version> Memory<P, V> {
     /// it's a total solution and version solving has succeeded.
     pub fn extract_solution(&self) -> Option<Map<P, V>> {
-        if self.assignments.values().all(|pa| pa.is_valid()) {
+        if self.assignments.iter().all(|(_, pa)| match pa {
```
If we embed the validation directly in `extract_solution` we might as well do a single pass on assignments. We can initialize an empty `decisions` hashmap, iterate in a for loop on assignments, grow the decisions until we reach the end (and return `Some(decisions)`) or a positive derivation (and return `None`).

Eh2406 replied:
I considered that. It felt like a close call between "one pass over assignments" vs "only allocates if needed". Now I see that it is called only when we expect it to work, so "one pass" is better, will fix.

mpizenberg commented on src/internal/incompatibility.rs:
```diff
@@ -225,16 +225,15 @@ impl<P: Package, V: Version> Incompatibility<P, V> {
+                    return Relation::Contradicted(package.clone(), incompat_term.clone());
                 }
-                term::Relation::Inconclusive => {
+                _ => {
```
Ok, if a package is not present, the intersection is the same as `any` and according to the rules of satisfactions, the relation would be inconclusive. It could also be satisfied if the incompatibility term was also `any` but we systematically remove those from incompatibilities so we're safe on that front. We might add a comment saying just that.

Eh2406 replied:
Will add

mpizenberg commented on src/internal/memory.rs:
```diff
@@ -36,11 +38,10 @@ impl<P: Package, V: Version> Memory<P, V> {
-            None => Term::any(),
-            Some(pa) => pa.assignment_intersection(),
-        }
+    pub fn term_intersection_for_package(&mut self, package: &P) -> Option<&Term<V>> {
```
Now returning a reference (in addition to option), ok that's better.

mpizenberg commented on src/internal/memory.rs:
```diff
@@ -106,69 +139,44 @@ impl<P: Package, V: Version> Memory<P, V> {
+                intersected,
+                not_intersected_yet,
+            } => {
+                for derivation in not_intersected_yet.drain(..) {
```
Dumb performance question. Is `drain` optimized to be as fast as a `clear` when all values are extracted? I've not tested it but went the other way the first time just from intuition it may not.

Eh2406 replied:
Now that you say it, `drain(..)` probably has to do some bookkeeping for panic safety that iter and clear do not, I will switch it back.

mpizenberg commented on src/solver.rs:
```diff
@@ -126,47 +126,49 @@ pub fn resolve<P: Package, V: Version>(
             .or_default()
             .insert(v.clone())
         {
+            // Retrieve that package dependencies.
```
that diff block is actually just moved code ok

mpizenberg commented on src/solver.rs:
```diff
@@ -126,47 +126,49 @@ pub fn resolve<P: Package, V: Version>(
-            ));
-        }
-        state.partial_solution.add_version(p, v, &dep_incompats);
+        state.partial_solution.add_decision(p, v);
```
So `add_version` is replaced by `add_decision`.

mpizenberg commented on src/internal/memory.rs:
```diff
@@ -53,27 +54,50 @@ impl<P: Package, V: Version> Memory<P, V> {
-            .entry(package)
-            .or_insert_with(PackageAssignments::new);
-        pa.derivations_not_intersected_yet.push(term);
+        let pa =
```
We might have a `PackageAssignments::init_with_derivation(term)` constructor

mpizenberg commented on src/internal/memory.rs:
```diff
@@ -53,27 +54,50 @@ impl<P: Package, V: Version> Memory<P, V> {
+            self.assignments
+                .entry(package)
+                .or_insert_with(|| PackageAssignments::Derivations {
+                    intersected: Term::any(),
```
If there is only one term, no need to put it in the `not_intersected_yet` vec, we can put it directly in `intersected`

mpizenberg commented on src/internal/memory.rs:
```diff
@@ -53,27 +54,50 @@ impl<P: Package, V: Version> Memory<P, V> {
+        match pa {
+            PackageAssignments::Decision((version, _)) => {
+                if cfg!(debug_assertions) {
+                    debug_assert!(term.contains(&version))
```
I wonder if you could even put `debug_assert!(false)`. I don't think we ever add a derivation for a package where a decision was already taken since we only add derivations for almost satisfied packages on incompatibilities (in unit propagation) and a package with a decision already cannot return `Inconclusive` I think

mpizenberg commented on src/internal/memory.rs:
```diff
@@ -53,27 +54,50 @@ impl<P: Package, V: Version> Memory<P, V> {
+            } => {
+                not_intersected_yet.push(term);
+            }
+        }
```
In sum, I'd do something like this. Except there is a clone there because of moved value that we should get rid of. I'm not sure exactly how.
```rust
    /// Add a derivation to a Memory.
    fn add_derivation(&mut self, package: P, term: Term<V>) {
        self.assignments
            .entry(package)
            .and_modify(|pa| match pa {
                PackageAssignments::Decision(_) => {
                    if cfg!(debug_assertions) {
                        debug_assert!(false)
                    }
                }
                PackageAssignments::Derivations {
                    intersected: _,
                    not_intersected_yet,
                } => {
                    not_intersected_yet.push(term.clone());
                }
            })
            .or_insert_with(|| PackageAssignments::Derivations {
                intersected: term,
                not_intersected_yet: vec![],
            });
    }
```

mpizenberg commented on src/internal/memory.rs:
```diff
@@ -53,27 +54,50 @@ impl<P: Package, V: Version> Memory<P, V> {
 
     /// Add a decision to a Memory.
```
```rust
    /// We can add the version to the partial solution as a decision
    /// if it doesn't produce any conflict with the new incompatibilities.
    /// In practice I think it can only produce a conflict if one of the dependencies
    /// (which are used to make the new incompatibilities)
    /// is already in the partial solution with an incompatible version.
```

mpizenberg replied:
That check that was done in the removed `add_version` was important. It prevents generating a future conflict and thus avoid a full pass of conflict resolution. That's the heart of what I meant when I said that removing the `remove_decision` function would mean a few changes in the algorithm. We have to have a way to prevent that decision been added if one of it's dependencies incompatibilities would become satisfied.

One option I can see is checking if the dependencies are almost satisfied before adding the decision. If that's the case, it is necessarily because the unsatisfied term is the decision we want to add. And in that case we shouldn't add it since it would mean that we will satisfy one of those incompatibilities.

mpizenberg replied:
> One option I can see is checking if the dependencies are almost satisfied before adding the decision

Actually this is wrong since the derivations intersection for the package we are adding could already match the decision term, and in such case the incompatibility could already be fully satisfied, not just almost satisfied. But since we've already validated the dependencies at that point, the correct (and most efficient) way is I think to use the dependency ranges directly. We could check that none of the `not(range)` terms for dependency ranges are satisfied by the partial solution before adding the decision. Because if any of them are satisfied, we know for sure that the incompatibility `{p: v, dep: not(range)}` will be satisfied once we add the decision for version `v`.

Eh2406 replied:
> avoid a full pass of conflict resolution.

The full pass will end up in the same place, so that is what I had in mind by calling this a "fast path". But if it is important there are a number of "simple" ways to add it back in. One is to redefine `remove_last_decision` as `bactrack(self.decision -1)`. Another is to "ask permission not forgiveness" in this case, and... it is hard to say in words but it worked so take a look.

Also now I think we only need to do this if the dependencies are new, as otherwise the incompatibilities are already accounted for by the previous call to `unit_propagation`.

Eh2406 committed:
refactor: suggestions

Eh2406 committed:
refactor: readd add_version asking permission not forgiveness

mpizenberg committed:
refactor: simplify debug assertions in add_decision

mpizenberg commented on src/internal/memory.rs:
```diff
@@ -53,27 +54,48 @@ impl<P: Package, V: Version> Memory<P, V> {
+                } => {
+                    not_intersected_yet.push(term);
+                }
+            },
```
Very elegant!

mpizenberg committed:
docs: add a debug_assert comment

mpizenberg committed:
refactor: remove &* in assignment_intersection

mpizenberg committed:
refactor: use iter().any() in add_version preventive check

mpizenberg reviewed (approved):
Regarding my commit removing `&*` was there a reason I'm not aware of for having those?
Last commit is a matter of taste I believe. I tend to avoid for loops if possible. Feel free to change back if you don't like the change.

aleksator committed:
docs: mention Term::any explicitly

aleksator committed:
refactor: remove early return in add_version()

Eh2406 commented:
Shit. The commit created by the squash does not follow the style. I can't force push to dev. So I don't know how to fix it. Sorry!

aleksator commented:
Interesting. That's what `bors` would prevent, right?

aleksator commented:
> Eh2406 merged commit 9f6b511 into dev 1 hour ago

It says 5 checks pass, what's the problem? :thinking:

aleksator commented:
Ah I see, no conventional commit prefix there. I'll see if I can fix it.

Eh2406 commented:
https://github.com/pubgrub-rs/pubgrub/commit/9f6b511b02f6c090d0e0105954f0f9f122a95056

Has a red x

On Sun, Nov 15, 2020, 11:03 AM Alex Tokarev <notifications@github.com>
wrote:

> Eh2406 merged commit 9f6b511
> <https://github.com/pubgrub-rs/pubgrub/commit/9f6b511b02f6c090d0e0105954f0f9f122a95056>
> into dev 1 hour ago
>
> It says 5 checks pass, what's the problem? 🤔
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/pubgrub-rs/pubgrub/pull/58#issuecomment-727593286>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA4JUQEK6R5DLERQW7LWKG3SP73TTANCNFSM4TG52NSA>
> .
>

aleksator commented:
I temporarily allowed force pushes and changed commit name. :man_shrugging:

Eh2406 commented:
Starting now.

Eh2406 commented:
Oops you did it already.

Eh2406 commented:
I tried to put the protections back, but I don't know if I got it correct. Sorry for messing things up.

aleksator commented:
> I tried to put the protections back, but I don't know if I got it correct. Sorry for messing things up.

I restored them after pushing, they should have been already fine :slightly_smiling_face: Don't worry! No big deal.

Eh2406 commented:
>  they should have been already fine

before I deleted it. 🤦 
You fixed it. Then I posted. Then I removed the protections. (dumly by delling, instead of unchecking the one that matters.) Then I noticed that you took care of it, and tried to put things back.

aleksator commented:
It's fine, relax! 😃

I've checked the settings for the `dev` branch, the only difference is we didn't have `Require status checks to pass before merging` selected before. I don't think turning it on even worked since no required checks were selected (there are CI jobs that you mark as required after choosing that option). I've unchecked the whole thing just to make it clearer.

In my understanding, what that option does is it requires the branch to be rebased on dev before you can use `squash merge` to run the branch through CI again. That's something we might want (or not). We can discuss it on Zulip if you feel like.
